src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -27888,6 +27888,66 @@ void Test1(Test? x1)
                 );
         }
 
+        [WorkItem(31018, "https://github.com/dotnet/roslyn/issues/31018")]
+        [Fact]
+        public void EventAssignment()
+        {
+            var source =
+@"#pragma warning disable 0067
+using System;
+class A<T> { }
+class B
+{
+    event Action<A<object?>> E;
+    static void M1()
+    {
+        var b1 = new B();
+        b1.E += F1; // 1
+        b1.E += F2; // 2
+        b1.E += F3;
+        b1.E += F4;
+    }
+    static void M2(Action<A<object>> f1, Action<A<object>?> f2, Action<A<object?>> f3, Action<A<object?>?> f4)
+    {
+        var b2 = new B();
+        b2.E += f1; // 3
+        b2.E += f2; // 4
+        b2.E += f3;
+        b2.E += f4;
+    }
+    static void M3()
+    {
+        var b3 = new B();
+        b3.E += (A<object> a) => { }; // 5
+        b3.E += (A<object>? a) => { }; // 6
+        b3.E += (A<object?> a) => { };
+        b3.E += (A<object?>? a) => { }; // 7
```

Comment:
This seems unfortunate, but given that this is already how variance works for lambdas/method groups, I don't see a way to change this without making overload resolution rules extremely convoluted and inconsistent.